fpdf/line_break.py

```diff
@@ -0,0 +1,237 @@
+
+    def add_character(self, character):
+        self.characters.append(character)
+
+    def trim(self, index):
+        self.characters = self.characters[:index]
+
+    @property
+    def string(self):
+        return "".join(self.characters)
+
+    def __eq__(self, other):
+        return (
+            self.characters == other.characters
+            and self.style == other.style
+            and self.underline == other.underline
+        )
+
+
+text_line = namedtuple(
+    "text_line",
+    ("fragments", "text_width", "number_of_spaces_between_words", "justify"),
+)
+
+space_hint = namedtuple(
+    "break_hint",
+    (
+        "original_fragment_index",
+        "original_character_index",
+        "current_line_fragment_index",
+        "current_line_character_index",
+        "width",
+        "number_of_spaces",
+    ),
+)
+
+hyphen_hint = namedtuple(
+    "hyphen_hint",
+    space_hint._fields
+    + (
+        "character_to_append",
+        "character_to_append_width",
+        "character_to_append_style",
+        "character_to_append_underline",
+    ),
+)
+
+
+class CurrentLine:
+    def __init__(self):
+        self.fragments = []
+        self.width = 0
+        self.number_of_spaces = 0
+
+        # automatic break hints
+        self.space_break_hint = None
+        self.hyphen_break_hint = None
+
+    def add_character(
+        self,
+        character,
+        character_width,
+        style,
+        underline,
+        original_fragment_index,
+        original_character_index,
+    ):
+        assert character != NEWLINE
+
+        if not self.fragments:
+            self.fragments.append(Fragment(style, underline))
+
+        elif (
+            style != self.fragments[-1].style
+            or underline != self.fragments[-1].underline
+        ):
+            self.fragments.append(Fragment(style, underline))
+        active_fragment = self.fragments[-1]
+
+        if character == SPACE:
+            self.space_break_hint = space_hint(
+                original_fragment_index,
+                original_character_index,
+                len(self.fragments),
+                len(active_fragment.characters),
+                self.width,
+                self.number_of_spaces,
+            )
+            self.number_of_spaces += 1
+        elif character == SOFT_HYPHEN:
+            self.hyphen_break_hint = hyphen_hint(
+                original_fragment_index,
+                original_character_index,
+                len(self.fragments),
+                len(active_fragment.characters),
+                self.width,
+                self.number_of_spaces,
+                HYPHEN,
+                character_width,
+                style,
+                underline,
+            )
+
+        if character != SOFT_HYPHEN:
+            self.width += character_width
+            active_fragment.characters.append(character)
+
+    def _apply_automatic_hint(self, break_hint):
```

Comment:
Could you please add a brief docstring line about what this method does?

Reply:
will do